packages/core/src/indexing/insertion-order.ts

```diff
@@ -177,11 +197,45 @@ export class InsertionOrder {
         base = base.where(converted.where)
       }
 
-      if (isReverseOrder) {
-        base = base.orderBy(reverseOrder(INSERTION_ORDER))
+      // Handle cursor if present
+      if (cursor != null) {
+        if (cursor.type === 'timestamp') {
+          // Paginate using the `created_at` field when no custom field ordering is provided
+          base = base.where((qb) => {
+            qb.where('created_at', isReverseOrder ? '<' : '>', cursor.value) // strict next value
+              .orWhere('created_at', '=', cursor.value) // or current value
+              .andWhere('stream_id', '>', cursor.id) // with stream ID tie-breaker
+          })
+        } else {
+          // Paginate using previous values of custom fields
+          for (const [key, prevValue] of Object.entries(cursor.value)) {
+            const field = contentKey(key)
+            const sign = getComparisonSign(query.sorting?.[key], isReverseOrder)
+            const value = JSON.stringify(prevValue)
+            base = base.where((qb) => {
+              qb.whereRaw(`${field} ${sign} ${value}`) // strict next value
+                .orWhereRaw(`${field} = ${value}`) // or current value
+                .andWhere('stream_id', '>', cursor.id) // with stream ID tie-breaker
```

Comment:
should we still use created_at as the first tiebreaker for the explicit sort keys and only fall back to stream_id in the rare event of a tie on created_at?

Reply:
With the logic here the cursor always contains the stream `id`, but either contains the `created_at` value or the custom field(s) value(s).
We could always store the `created_at` value in the cursor, but we need to define the behavior more I think: do we also need to order by `created_at` in any way? It would add complexity to the logic and it's not a "fully reliable" tie-breaker as we'd still need to use the `stream_id` so I'd rather just use the `stream_id` if possible?
Also always adding the `created_at` field to the cursor makes the cursor longer. I don't think it's a big issue but seems a bit wasteful to make the payloads larger than they need to be.

Reply:
yeah, I was thinking we would add `created_at` to the sort as well, as it's a slightly more meaningful thing to sort on than `stream_id`.  Really ideally the default sort would be `last_anchored_at` first, then `created_at`, then `stream_id`.  But yes, that does keep inflating the size of the cursor we pass back and forth between the client and server.

But I don't feel super strongly about this, we can always change later if users ask for different behavior.